spec/requests/users_controller_spec.rb

```diff
@@ -1,51 +1,2402 @@
+          user.expects(:enqueue_welcome_message).with('welcome_user').never
+
+          put "/u/activate-account/#{token}"
+        end
+      end
+
+      context "honeypot" do
+        it "raises an error if the honeypot is invalid" do
+          UsersController.any_instance.stubs(:honeypot_or_challenge_fails?).returns(true)
+          put "/u/activate-account/#{token}"
+          expect(response).not_to be_success
+        end
+      end
+
+      context 'response' do
+        before do
+          Guardian.any_instance.expects(:can_access_forum?).returns(true)
+          EmailToken.expects(:confirm).with("#{token}").returns(user)
+        end
+
+        it 'correctly logs on user' do
+          events = DiscourseEvent.track_events do
+            put "/u/activate-account/#{token}"
+          end
+
+          expect(events.map { |event| event[:event_name] }).to include(
+            :user_logged_in, :user_first_logged_in
+          )
+
+          expect(response).to be_success
+          expect(flash[:error]).to be_blank
+          expect(session[:current_user_id]).to be_present
+
+          expect(response).to be_success
+
+          expect(CGI.unescapeHTML(response.body))
+            .to_not include(I18n.t('activation.approval_required'))
+        end
+      end
+
+      context 'user is not approved' do
+        before do
+          SiteSetting.must_approve_users = true
+          EmailToken.expects(:confirm).with("#{token}").returns(user)
+          put "/u/activate-account/#{token}"
+        end
+
+        it 'should return the right response' do
+          expect(response).to be_success
+
+          expect(CGI.unescapeHTML(response.body))
+            .to include(I18n.t('activation.approval_required'))
+
+          expect(response.body).to_not have_tag(:script, with: {
+            src: '/assets/application.js'
+          })
+
+          expect(flash[:error]).to be_blank
+          expect(session[:current_user_id]).to be_blank
+        end
+      end
+    end
+  end
+
+  describe '#perform_account_activation' do
+    describe 'when cookies contains a destination URL' do
+      let(:token) { 'asdadwewq' }
+      let(:user) { Fabricate(:user) }
+
+      before do
+        UsersController.any_instance.stubs(:honeypot_or_challenge_fails?).returns(false)
+        EmailToken.expects(:confirm).with(token).returns(user)
+      end
+
+      it 'should redirect to the URL' do
+        destination_url = 'http://thisisasite.com/somepath'
+        cookies[:destination_url] = destination_url
+
+        put "/u/activate-account/#{token}"
+
+        expect(response).to redirect_to(destination_url)
+      end
+    end
+  end
+
+  describe '#password_reset' do
+    let(:user) { Fabricate(:user) }
+    let(:token) { SecureRandom.hex }
+
+    context "you can view it even if login is required" do
+      it "returns success" do
+        SiteSetting.login_required = true
+        get "/u/password-reset/#{token}"
+        expect(response).to be_success
```

Comment:
Feels like we're missing a couple of assertions here. Is there a response body that we're expecting or should something in the DB change?